src/usr/local/www/firewall_rules_edit.php

```diff
@@ -1287,14 +1299,50 @@ function build_if_list() {
 	)
 ))->setHelp('Choose which IP protocol this rule should match.');
 
-$group = new Form_Group("ICMP Subtypes");
-$group->add(new Form_Select(
-	'icmptype',
-	'ICMP subtypes',
-	((isset($pconfig['icmptype']) && strlen($pconfig['icmptype']) > 0) ? explode(',', $pconfig['icmptype']) : 'any'),
-	isset($icmplookup[$pconfig['ipprotocol']]) ? $icmplookup[$pconfig['ipprotocol']]['icmptypes'] : array('any' => gettext('any')),
-	true
-))->setHelp('<div id="icmptype_help">' . (isset($icmplookup[$pconfig['ipprotocol']]) ? $icmplookup[$pconfig['ipprotocol']]['helpmsg'] : '') . '</div>');
+$group = new Form_Group('ICMP Subtypes'); 
+
+// generate ICMPTYPE checkboxes
+// If no existing selection, enable all.  Any not available under selected IP protocol will be hidden, & ignored on submit.
+// This probably works better for UI than using array_intersect to exclude them here
+$sel = (isset($pconfig['icmptype']) && strlen($pconfig['icmptype']) > 0) ? explode(',', $pconfig['icmptype']) : array_keys($icmptypes);
+$html = '';
+foreach ($icmptypes as $type => $data) {
+	$chkbox = new Form_MultiCheckbox(
+		"icmptype_{$type}",
+		null,
+		$data['descrip'],
+		in_array($type, $sel) ? true : false,
+		$type
+	);
+	// manually edit bootstrap output: it sets name = id and we don't want that (we want name to reference an array)
+	$chkbox = str_replace("name=\"icmptype_{$type}\"", "name=\"icmptype_selection[]\"", $chkbox);
+	//wrap checkboxes with show/hide classes and a bit of spacing
+	$extraclasses = 'icmptype_chkbox' . 
+		($data['valid4'] ? ' icmptype_valid_inet' : '') .
+		($data['valid6'] ? ' icmptype_valid_inet6' : '') .
+		(($data['valid4'] && $data['valid6']) ? ' icmptype_valid_inet46' : '');
+	$html .= "<span class='{$extraclasses}' style='padding-right:10px'>{$chkbox}</span>\n";
+}
+
+$buttonall = (new Form_Button(
+	'btn_icmp_selectall',
+	'Select all',
+	null,
+	''
```

Comment:
Done. Also same for "deselectall" checkbox, and removed a completely useless `? true:false' clause a bit below them.